projects/driver_controls/src/digital_io.c

```diff
@@ -0,0 +1,121 @@
+#include <stddef.h>
+
+#include "gpio_it.h"
+#include "digital_io.h"
+#include "input_event.h"
+
+// Digital device identifiers
+typedef enum {
+  DIGITAL_IO_POWER_SWITCH = 0,
+  DIGITAL_IO_DIRECTION_SELECTOR,
+  DIGITAL_IO_CRUISE_CONTROL,
+  DIGITAL_IO_CRUISE_CONTROL_INC,
+  DIGITAL_IO_CRUISE_CONTROL_DEC,
+  DIGITAL_IO_TURN_SIGNAL,
+  DIGITAL_IO_HAZARD_LIGHT,
+  DIGITAL_IO_MECHANICAL_BRAKE,
+  NUM_DIGITAL_IO_INPUTS
+} DigitalIODevices;
+
+// Store the id of the device as well as the id of the event the device raises
+typedef struct DigitalIOData {
+  DigitalIODevices id;
+  InputEvent event;
+} DigitalIOData;
+
+// Configuration settings for the digital pins
+typedef struct DigitalIOSettings {
+  GPIOAddress address;
+  InterruptEdge edge;
+  DigitalIOData *data;
+} DigitalIOSettings;
+
+// Index the objects using their respective pins
+static DigitalIOData s_input_data[] = {
+  [0] = { .id = DIGITAL_IO_POWER_SWITCH, .event = INPUT_EVENT_POWER },
+  [2] = { .id = DIGITAL_IO_DIRECTION_SELECTOR, .event = INPUT_EVENT_DIRECTION_SELECTOR_DRIVE },
+  [3] = { .id = DIGITAL_IO_DIRECTION_SELECTOR, .event = INPUT_EVENT_DIRECTION_SELECTOR_REVERSE },
+  [4] = { .id = DIGITAL_IO_CRUISE_CONTROL, .event = INPUT_EVENT_CRUISE_CONTROL },
+  [5] = { .id = DIGITAL_IO_CRUISE_CONTROL_INC, .event = INPUT_EVENT_CRUISE_CONTROL_INC },
+  [6] = { .id = DIGITAL_IO_CRUISE_CONTROL_DEC, .event = INPUT_EVENT_CRUISE_CONTROL_DEC },
+  [7] = { .id = DIGITAL_IO_TURN_SIGNAL, .event = INPUT_EVENT_TURN_SIGNAL_RIGHT },
+  [8] = { .id = DIGITAL_IO_TURN_SIGNAL, .event = INPUT_EVENT_TURN_SIGNAL_LEFT },
+  [9] = { .id = DIGITAL_IO_HAZARD_LIGHT, .event = INPUT_EVENT_HAZARD_LIGHT },
+  [10] = { .id = DIGITAL_IO_MECHANICAL_BRAKE, .event = INPUT_EVENT_MECHANICAL_BRAKE }
+};
+
+// Genarate the event based on the identity of the triggering device
+static void prv_get_event(DigitalIOData *digital_io_data, Event *e, GPIOState state) {
```

Comment:
Just so you're aware, most IO is active-low. This means that 0V/GND = high, digital 1 while 3.3V/VCC = low, digital 0.